### ISSUE TITLE
Can the styling work out what document#head to insert itself into?

### ISSUE THREAD

**Issue by sebinsua** [emotion-js/emotion#560]
I have an interesting use-case which from what I can see doesn't seem to be easily supported.

I have an enterprise React application which may or may not be rendered into either an iframe or a separate window (via [`openfin#Window`](http://cdn.openfin.co/jsdocs/beta/fin.desktop.Window.html#.getCurrent)).

My understanding is that, the styling will not work unless the `<style>` tags are placed into the correct window. I can see that it's [possible to pass in a `container`](https://github.com/emotion-js/emotion/blob/master/packages/create-emotion/src/sheet.js#L53-L55), however I also see that by default [`createEmotion` is called at import](https://github.com/emotion-js/emotion/blob/master/packages/emotion/src/index.js#L17) by `emotion` and `react-emotion`. In order to load the styles into the correct window, I would need to not use those modules but instead use the raw `create-emotion` and create my own wrappers that pass in the correct container.

I might be able to get this to work by setting the `document#head` based on either [`openfin#Window#getCurrent`](http://cdn.openfin.co/jsdocs/beta/fin.desktop.Window.html#.getCurrent), the `document` or the the `iframe`. However, this must happen at runtime and in any case I'm not in charge of when these windows are created, or how they will be configured so it might not be possible to know how to select the correct `document`s.

What I normally do is use a `ref` within my React code to get a DOM element and then check this for the [`element#ownerDocument`](https://developer.mozilla.org/en-US/docs/Web/API/Node/ownerDocument) to find out the real `document`. However, this pattern will not work in this case since the `sheet` will have already been created at import-time, and any styles created with the `css` template literal will have already added their respective rules to the wrong `document#head`.

I'm wondering whether you can think of any patterns which would help me to get this to work without too much boilerplate or complication? I would like to be able to write normal shareable components which do not need to know about my environment.

Is there be a way of configuring the underlying `sheet` which the styles are attached to at runtime and potentially choosing to assign styles to a new `document` element after they were originally inserted? In that case I could call this method when rendering my React Application and could configure the styles by passing in a `refElement.ownerDocument.head`.

Sorry for the rambling message. Hopefully you understand the question/feature request.

**Comment by mitchellhamilton** [comment]
Doing something like this isn't really possible with emotion's current architecture. I'm currently working on something that would make it possible to solve this problem though in https://github.com/emotion-js/next.

**Comment by herrstucki** [comment]
For reference, I have a similar problem: https://github.com/interactivethings/catalog/issues/376

**Comment by sebinsua** [comment]
I came up with an awful hack for my use-case.

```ts
import React, { Component } from 'react';
import { sheet } from 'emotion';

export class EmotionSheetProvider extends Component {

  setContainer = (ref) => (this.container = ref);

  componentDidMount() {
    // This method is called after a render when we know the container has been set.
    // At this point we have a <div> element that we can get the owner document off,
    // in order that the code runs on the correct document window.
    if (this.container) {
      const ownerDocument = this.container.ownerDocument!;
      const head = ownerDocument.getElementsByTagName('head')[0];

      // This is a hack.
      // If emotion is rendered into an <iframe>, it needs its container
      // to be the <head> of the frame which it is rendering into,
      // instead of the parent frame in which the JavaScript is executed.
      // See: https://github.com/emotion-js/emotion/issues/560
      const emotionSheet = sheet;
      Object.defineProperty(emotionSheet.opts, 'container', {
        value: head
      });
      if ('__SECRET_EMOTION__' in window) {
        const secretEmotionSheet = window.__SECRET_EMOTION__.sheet;
        Object.defineProperty(secretEmotionSheet.opts, 'container', {
          value: head
        });
      }
    }
  }

  render() {
    const { children } = this.props;
    return (
      <div className="EmotionSheetProvider" ref={this.setContainer}>
        {children}
      </div>
    );
  }
}

export default EmotionSheetProvider;
```

It's not possible for me to use `createEmotion` directly, because the container my application will be placed into is only known at runtime. The parent application is OpenFin/Electron and can place it into new windows or new frames and I need for it to just work.

> Doing something like this isn't really possible with emotion's current architecture. I'm currently working on something that would make it possible to solve this problem though in https://github.com/emotion-js/next.

@mitchellhamilton Glad to see you're working on this.

I think my big remaining worry, is what I will do when there is a single instance of emotion in a parent window with all of the JavaScript, but 10s of child windows opened up, which should each have their own `<style>` tags within their own `<head>` tags. I'm worried that this won't be possible, because there is only one `container`.

**Comment by sebinsua** [comment]
Just to add, I think the only css-in-js framework to have handled this correctly so far is `jss`.

They have a [`<JssProvider jss={jssInstance} />`](https://github.com/cssinjs/react-jss#custom-setup) component, as well as a [`jssInstance#insertionPoint`](https://github.com/cssinjs/jss/blob/master/docs/js-api.md#setup-jss-instance).

**Comment by erquhart** [comment]
Would be really cool to see this solved in Emotion. Styled Components has a similar approach to the one JSS uses, a `StylesheetManager` component.

**Comment by Andarist** [comment]
@erquhart it's possible with custom emotion instances (you can check out how to create 1 [here](https://github.com/emotion-js/emotion/issues/712#issuecomment-395912093))

And it will become even easier in upcoming emotion@10 - see [this](https://github.com/emotion-js/emotion/issues/760#issuecomment-404353706)

**Comment by erquhart** [comment]
The rough part with instances is making them available to other packages within our monorepo. We can create it in the core package, just no clean way to pass it around. I do have my eyes on v10, is it even remotely safe to use yet? I'm okay with early adoption :)

**Comment by peduarte** [comment]
@mitchellhamilton hey there, i'm quite new to emotion, but isn't there a less low level way to do this?
Currently with SC, I have:
```
<StyleSheetManager target={frameContext.document.head} ...>
```

Thanks!

**Comment by erquhart** [comment]
Agreed, that's a whole lot of integration code in the linked sandbox. Is it fair to say this issue should remain open until emotion handles most/all of this? Definitely awesome that v10 opens the door! Would just like to see built in support.
